wave_utils/src/pose_cov_comp.cpp

```diff
@@ -0,0 +1,514 @@
+    p27 = p2.getPoseQuaternion();
+
+    p16 << p1.getPosition(), quatToYPR(p17.block<4, 1>(3, 0));
+    p26 << p2.getPosition(), quatToYPR(p27.block<4, 1>(3, 0));
+
+    // Equation (5.3)
+    Matrix6x7 jacobian_p7_to_p6 = jacobian_p7_to_p6_wrt_p(pR7);
+    Matrix7x7 jacobian_p7_p7_composition =
+      jacobian_p7_p7_Composition_wrt_p1(p17, p27);
+    Matrix7x6 jacobian_p6_to_p7 = jacobian_p6_to_p7_wrt_p(p16);
+    Matrix6x6 dfpc_dp =
+      jacobian_p7_to_p6 * jacobian_p7_p7_composition * jacobian_p6_to_p7;
+
+    // Equation (5.4)
+    jacobian_p7_p7_composition = jacobian_p7_p7_Composition_wrt_p2(p17, p27);
+    jacobian_p6_to_p7 = jacobian_p6_to_p7_wrt_p(p26);
+    Matrix6x6 dfpc_dq =
+      jacobian_p7_to_p6 * jacobian_p7_p7_composition * jacobian_p6_to_p7;
+
+    // Equation (5.2)
+    // putting everything together
+    Matrix6x6 final_cov = dfpc_dp * cov_p1 * dfpc_dp.transpose() +
+                          dfpc_dq * cov_p2 * dfpc_dq.transpose();
+
+    r.covariance = final_cov;
+
+    return r;
+}
+
+Vector4 normalizeQuat(const Vector4 q) {
+    Vector4 norm_q;
+    double norm_coeff;
+    double qr = q(0), qx = q(1), qy = q(2), qz = q(3);
+
+    norm_coeff = 1 / sqrt(qr * qr + qx * qx + qy * qy + qz * qz);
+    norm_q = norm_coeff * q;
+    return norm_q;
+}
+
+Vector3 quatToYPR(const Vector4 &q) {
+    Vector3 v;
+    double qr = q(0), qx = q(1), qy = q(2), qz = q(3);
+    double delta = qr * qy - qx * qz;
+    double roll, pitch, yaw;
+
+    if (fabs(delta - 0.5) < 1e-6) {
+        yaw = -2 * atan2(qx, qr);
+        pitch = M_PI / 2;
+        roll = 0;
+
+    } else if (fabs(delta + 0.5) < 1e-6) {
+        yaw = 2 * atan2(qx, qr);
+        pitch = -M_PI / 2;
+        roll = 0;
+    } else {
+        yaw = atan2(2 * (qr * qz + qx * qy), (1 - 2 * (qy * qy + qz * qz)));
+        pitch = asin(2 * delta);
+        roll = atan2(2 * (qr * qx + qy * qz), (1 - 2 * (qx * qx + qy * qy)));
+    }
+
+    v(0) = yaw;
+    v(1) = pitch;
+    v(2) = roll;
+
+    return v;
+}
+
+Vector4 yprToQuat(const Vector3 &ypr) {
+    Vector4 q;
+    double qr, qx, qy, qz;
+    double half_y = ypr(0) / 2, half_p = ypr(1) / 2, half_r = ypr(2) / 2;
+    double cy = cos(half_y), cp = cos(half_p), cr = cos(half_r),
+           sy = sin(half_y), sp = sin(half_p), sr = sin(half_r);
+
+    qr = cr * cp * cy + sr * sp * sy;
+    qx = sr * cp * cy - cr * sp * sy;
+    qy = cr * sp * cy + sr * cp * sy;
+    qz = cr * cp * sy - sr * sp * cy;
+
+    q << qr, qx, qy, qz;
+
+    return q;
+}
+
+Matrix3x3 yprToRotMatrix(const Vector3 &ypr) {
+    double y = ypr(0), p = ypr(1), r = ypr(2);
+    Matrix3x3 rotMatrix;
+    double cy = cos(y), cp = cos(p), cr = cos(r), sy = sin(y), sp = sin(p),
+           sr = sin(r);
+
+    rotMatrix << cy * cp, cy * sp * sr - sy * cr, cy * sp * cr + sy * sr,
+      sy * cp, sy * sp * sr + cy * cr, sy * sp * cr - cy * sr, -sp, cp * sr,
+      cp * cr;
+
+    return rotMatrix;
+}
+
+Vector3 rotMatrixToYPR(const Matrix3x3 &p) {
```

Comment:
So some of these methods are ones that we've already implemented elsewhere in libwave. The `geometry` section as a whole is going through a bit of an overhaul anyways thanks to @leokoppel, so perhaps those methods become obsolete.

Reply:
https://github.com/wavelab/libwave/blob/master/wave_geometry/src/rotation.cpp

Reply:
Personally I feel that these should be in utils, as opposed to geometry.

Reply:
I took these functions straight from the paper; I was having some issues using Eigen to do the conversions for me.

Reply:
oh, no it's fine that you have your own functions for this. I just mean that we may already have these functions elsewhere in libwave - just double check to make sure there isn't any repetition

Reply:
For now, I've put them in their own namespace `wave::pose_comp` for clarification that these functions are used for pose composition only.

Reply:
Awesome, sounds good. Also @jskhu if you don't mind just adding a comment for the rationale that would be awesome. Just a quick sentence saying that this is slightly different than the other implementation in libwave and that this was used as it came straight from the paper.